### ISSUE TITLE
improve the picking code by implementing parallel execution

### ISSUE THREAD

**Issue by niketchhajed** [GeoscienceAustralia/hiperseis#86]
Currently, we are getting good results from the picking that ran on the temporary surveys like skippy, etc However, so far we were focusing on the effectiveness of the picks extracted, not on the efficiency. This task is to track the implementation of parallel execution of the picking codebase, both in terms of events in parallel as well as parallelism in trying the different permutations and combinations of the filtering frequency bands and picking parameters.

**Comment by niketchhajed** [comment]
Work done so far:
1. Wrote the parallel execution of picking based on multiple stations.
2. Tested (functionally and performance wose) and debugged the code.
3. Discovered that python2 has inherent limitations with multiprocessing library with the way fork system call based implementation.
4. Converted the script to python3 which has forkserver system call based multiprocessing implementation.
5. Tested again(functionally and performance wise) and observed that it gives much better performance but still not a linear scaling with the number of cores.
6. Checked in the code.
7. Had a chat with AWS support team on what EC2 VM instance types would be most optimal for our purposes.
8. Finally came up with a strategy: the target system would take the input in the form of a folder containing events pertaining to a given known time period.  The parallel execution script would take the folder as the input and output the extracted picks/origins/events into another given output folder.

Next steps:
1. Experiment with different high config EC2 instance types that are compute optimized.
2. Benchmark the performance of the new code base against a sample data set and compare apples-apples with the previous EC2 instance type and previous code base and arrive at the performance gain.

**Comment by niketchhajed** [comment]
After implementing the multiprocessing version of parallel picking and discussing the results with @BabakHejrani, it was apparent that mere 2 to 3 times of efficiency gain is not going to make the required difference in terms of turn around time. MPI on NCI was discussed as one of the options.

The first draft of the MPI version of parallel picking targeted for NCI is ready now and is functionally tested for a small dataset.

It is capable of running the picking based on the number of stations to be picked for a given event. NCI allows more than 256 npus to be used and our requirement lies within this range. The code needs to be tested for a bigger dataset and the benchmarking needs to be done for a dataset of commensurate size and the results need to be compared.

**Comment by zhang01GA** [comment]
no longer relevant
